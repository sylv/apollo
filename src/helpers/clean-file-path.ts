import path from "path";
import { ALL_EXTENSIONS, SUBTITLE_EXTENSIONS } from "../constants";
import { stripSpacePlaceholders } from "./strip-space-placeholders";

const EXPLICITLY_STRIP_PARTS = new Set([
  "tv shows",
  "videos",
  "movies",
  "shows",
  "series",
  "seasons",
  "import",
  "clean",
  "data",
  "library",
  "media",
  "mnt",
  "downloads",
  "active",
  "complete",
  "incomplete",
  "downloading",
  "torrents",
  "torrent",
]);

const IGNORE_PATH_PART_REGEX = /^(?:[A-Z]:|[a-z]+)$/i;
const EXCLUDE_BLACKLIST_REGEX = [
  /^lore$/i,
  /^histories(( and| &) lore)?$/i,
  /sample|^trailer(?! park)/i, // "excludes "trailer park boys"
  /^behind.the.scenes$/i,
  /^deleted.and.extended.scenes$/i,
  /^deleted.scenes$/i,
  /^extras?$/i,
  /^featurettes$/i,
  /^interviews$/i,
  /^scenes$/i,
  /^shorts$/i,
];

/**
 * Clean the file path and replace dots or underscores with spaces when necessary.
 * @returns undefined if the path is blacklisted.
 */
export function cleanFilePath(filePath: string): string | undefined {
  const filePathParts = filePath.split(/\/|\\/g);
  const cleanedPathParts: string[] = [];
  let hitUnignoredPart = false;

  for (let i = 0; i < filePathParts.length; i++) {
    const filePathPart = filePathParts[i];
    const lower = filePathPart.toLowerCase();
    if (EXPLICITLY_STRIP_PARTS.has(lower)) {
      continue;
    }

    const isLastPart = i === filePathParts.length - 1;
    const ext = isLastPart ? ALL_EXTENSIONS.find((ext) => lower.endsWith(ext)) : null;
    const withoutExtension = ext ? filePathPart.slice(0, -ext.length) : filePathPart;
    const cleanPathPartWithoutExt = stripSpacePlaceholders(withoutExtension);
    const cleanPathPart = ext ? `${cleanPathPartWithoutExt}${ext}` : cleanPathPartWithoutExt;
    // we have to check the blacklist first or else something like "Trailers" would be discarded
    // before we check the blacklist
    if (EXCLUDE_BLACKLIST_REGEX.some((p) => p.test(cleanPathPart))) return;
    // only once it passes the blacklist can we do this check
    if (filePathPart.length <= 2 || (IGNORE_PATH_PART_REGEX.test(filePathPart) && !hitUnignoredPart)) continue;
    hitUnignoredPart = true;
    cleanedPathParts.push(cleanPathPart);
  }

  return cleanedPathParts.join("/");
}
