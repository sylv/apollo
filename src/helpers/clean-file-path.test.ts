import { expect, it } from "vitest";
import { cleanFilePath } from "./clean-file-path";

const title = `Avengers Infinity War (2018) [WEBRip] [1080p] [YTS.AM].mp4`;
const directory = `Bobs Burgers`;

it("should clean file paths", () => {
  expect(cleanFilePath(`/mnt/data/tv shows/${title}`)).toBe(title);
  expect(cleanFilePath(`/mnt/data/tv shows/${directory}/${title}`)).toBe(`${directory}/${title}`);
});

it("should normalize path separators", () => {
  expect(cleanFilePath(`C:\\mnt\\data\\${directory}\\${title}`)).toBe(`${directory}/${title}`);
});

it("should not strip boring parts after interesting parts", () => {
  expect(cleanFilePath("/mnt/data/tv shows/Test file name (2016)/French.srt")).toBe("Test file name (2016)/French.srt");
  expect(
    cleanFilePath("/mnt/Downloads-2/import/Менталист/The.Mentalist.s07.WEB-DL.720p.Rus.Eng/The.Mentalist.s07e12-13.WEB-DL.720p.Rus.Eng.mkv")
  ).toMatchSnapshot();
});

it("should maintain file extensions", () => {
  expect(cleanFilePath("/mnt/data/tv shows/Test file name.srt")).toBe("Test file name.srt");
  expect(cleanFilePath("/mnt/data/tv shows/Test.file.name.srt")).toBe("Test file name.srt");

  // maintaining extensions sometimes thought the extension was ".264-[TestGroup]" for this incorrectly,
  // so now we maintain known extensions only.
  expect(cleanFilePath("Mythbusters.S2003.576p.PAL.AUS.AMZN.WEB-DL.DD+2.0.H.264-[TestGroup]")).toBe(
    "Mythbusters S2003 576p PAL AUS AMZN WEB-DL DD+2 0 H 264-[TestGroup]"
  );
});
