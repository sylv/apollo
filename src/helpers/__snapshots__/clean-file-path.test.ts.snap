// Vitest Snapshot v1

exports[`should not strip boring parts after interesting parts 1`] = `"Downloads-2/Менталист/The Mentalist s07 WEB-DL 720p Rus Eng/The Mentalist s07e12-13 WEB-DL 720p Rus Eng.mkv"`;
